File: HackerNewsAndroid/app/src/main/java/com/nurtrino/hackernews/ui/Theme.kt
package com.nurtrino.hackernews.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.isUnspecified
import com.nurtrino.hackernews.data.AppearanceMode
import com.nurtrino.hackernews.data.TextScale

/** The app icon's orange. */
val HnOrange = Color(0xFFFC5403)

/**
 * Per-depth rail colours for comment threads, and the source monograms.
 * Chosen to stay legible on both the light and dark surfaces.
 */
val ThreadPalette = listOf(
    HnOrange,
    Color(0xFF4C9EEB),
    Color(0xFF5CBD7A),
    Color(0xFFD971B8),
    Color(0xFF9985EB),
    Color(0xFFF0B83D),
    Color(0xFF66C7C7),
)

fun threadColor(depth: Int): Color = ThreadPalette[(if (depth < 0) -depth else depth) % ThreadPalette.size]

/** Stable colour for a string, so a domain always looks the same. */
fun stableColor(seed: String): Color {
    var hash = 5381L
    for (byte in seed.encodeToByteArray()) hash = hash * 33 + byte
    val index = ((hash % ThreadPalette.size) + ThreadPalette.size) % ThreadPalette.size
    return ThreadPalette[index.toInt()]
}

private val LightColors = lightColorScheme(
    primary = HnOrange,
    secondary = HnOrange,
    tertiary = HnOrange,
)

private val DarkColors = darkColorScheme(
    primary = HnOrange,
    secondary = HnOrange,
    tertiary = HnOrange,
)

object Metrics {
    const val MAX_INDENT_DEPTH = 8
}

@Composable
fun HackerNewsTheme(
    appearance: AppearanceMode,
    textScale: TextScale,
    content: @Composable () -> Unit,
) {
    val dark = when (appearance) {
        AppearanceMode.SYSTEM -> isSystemInDarkTheme()
        AppearanceMode.LIGHT -> false
        AppearanceMode.DARK -> true
    }

    MaterialTheme(
        colorScheme = if (dark) DarkColors else LightColors,
        typography = scaledTypography(textScale.factor),
        content = content,
    )
}

/** Applies the Settings text-size multiplier across the whole type scale. */
private fun scaledTypography(factor: Float): Typography {
    if (factor == 1.0f) return Typography()
    val base = Typography()

    fun TextStyle.scaled(): TextStyle = copy(
        fontSize = fontSize.scale(factor),
        lineHeight = lineHeight.scale(factor),
    )

    return Typography(
        displayLarge = base.displayLarge.scaled(),
        displayMedium = base.displayMedium.scaled(),
        displaySmall = base.displaySmall.scaled(),
        headlineLarge = base.headlineLarge.scaled(),
        headlineMedium = base.headlineMedium.scaled(),
        headlineSmall = base.headlineSmall.scaled(),
        titleLarge = base.titleLarge.scaled(),
        titleMedium = base.titleMedium.scaled(),
        titleSmall = base.titleSmall.scaled(),
        bodyLarge = base.bodyLarge.scaled(),
        bodyMedium = base.bodyMedium.scaled(),
        bodySmall = base.bodySmall.scaled(),
        labelLarge = base.labelLarge.scaled(),
        labelMedium = base.labelMedium.scaled(),
        labelSmall = base.labelSmall.scaled(),
    )
}

private fun TextUnit.scale(factor: Float): TextUnit =
    if (isUnspecified) this else this * factor

File: HackerNews/Support/Theme.swift
import SwiftUI

extension Color {
    /// Y Combinator orange, #FF6600.
    static let hnOrange = Color(red: 1.0, green: 0.4, blue: 0.0)

    /// Palette used for thread depth bars and source monograms. Chosen to stay
    /// legible against both the light and dark system backgrounds.
    static let threadPalette: [Color] = [
        .hnOrange,
        Color(red: 0.30, green: 0.62, blue: 0.92),
        Color(red: 0.36, green: 0.74, blue: 0.48),
        Color(red: 0.85, green: 0.44, blue: 0.72),
        Color(red: 0.60, green: 0.52, blue: 0.92),
        Color(red: 0.94, green: 0.72, blue: 0.24),
        Color(red: 0.40, green: 0.78, blue: 0.78),
    ]

    static func threadColor(depth: Int) -> Color {
        threadPalette[abs(depth) % threadPalette.count]
    }

    /// Stable colour for a string, so a given domain always looks the same.
    static func stableColor(for string: String) -> Color {
        var hash: UInt64 = 5_381
        for byte in string.utf8 {
            hash = (hash &* 33) &+ UInt64(byte)
        }
        return threadPalette[Int(hash % UInt64(threadPalette.count))]
    }
}

enum Metrics {
    static let rowSpacing: CGFloat = 6
    static let threadBarWidth: CGFloat = 2
    static let threadIndent: CGFloat = 11
    /// Beyond this the indentation eats the screen, so depth stops widening.
    static let maxIndentDepth = 8
}
